### PR TITLE
python 2.7 Unicode compatible, with test

### PR DESCRIPTION
**Pull request by niconoe**
To make life of the maintainer easier, I just grouped in this PR the fix for #79 by @dzhuang with the regression test by @tiumgraham.

### CODE REVIEW

**Comment by adamchainz** [comment]
* @timgraham :)

**Comment by dzhuang** [comment]
Thanks, but why it failed checks?

File: memcache.py [hunk 1]
```diff
@@ -1523,4 +1528,4 @@ def _doctest():
         sys.exit(1)
 
 
-# vim: ts=4 sw=4 et :
+# vim: ts=4 sw=4 et :
```
Comment (timgraham) [inline memcache.py]: revert the newline removal

Comment (adamchainz) [inline memcache.py]: this is missing a new line which makes it an invalid file in POSIX, that's why it failed travis on python 2.6

File: tests/test_memcache.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
+
```
Comment (timgraham) [inline tests/test_memcache.py]: no blank line needed

File: memcache.py [hunk 3]
```diff
@@ -1266,11 +1272,10 @@ def _recv_value(self, server, flags, rlen):
             flags &= ~Client._FLAG_COMPRESSED
 
         if flags == 0:
-            # Bare string
-            if six.PY3:
-                val = buf.decode('utf8')
-            else:
+            # Bare bytes
                 val = buf
```
Comment (timgraham) [inline memcache.py]: deindent this line

File: tests/test_memcache.py [hunk 4]
```diff
@@ -137,6 +139,14 @@ def test_unicode_key(self):
         value = self.mc.get(key)
         self.assertEqual(value, 5)
 
+    def test_unicode_value(self):
+        key = 'key'
+        value = six.u('Iñtërnâtiônàlizætiøn2')
+
```
Comment (timgraham) [inline tests/test_memcache.py]: I'd omit the blank line.

File: memcache.py (hunk 1 above)
Comment (niconoe) [inline memcache.py]: Nice catch, I was confused!

**Comment by timgraham** [comment]
Does it also fix #80?

File: tests/test_memcache.py [hunk 5]
```diff
@@ -137,6 +138,37 @@ def test_unicode_key(self):
...
+    def test_unicode_value(self):
+        key = 'key'
+        value = six.u('Iñtërnâtiônàlizætiøn2')
+        self.mc.set(key, value)
+        cached_value = self.mc.get(key)
+        self.assertEqual(value, cached_value)
+
+    def test_binary_string(self):
+        # Binary strings should be cacheable
+        from zlib import compress, decompress
```
Comment (timgraham) [inline tests/test_memcache.py]: I'd `import zlib` at the top of the file and use `zlib.compress()`, etc.

File: tests/test_memcache.py [hunk 6]
```diff
@@ -137,6 +138,37 @@ def test_unicode_key(self):
...
 
+    def test_unicode_value(self):
+        key = 'key'
+        value = six.u('Iñtërnâtiônàlizætiøn2')
+        self.mc.set(key, value)
+        cached_value = self.mc.get(key)
+        self.assertEqual(value, cached_value)
+
+    def test_binary_string(self):
+        # Binary strings should be cacheable
```
Comment (timgraham) [inline tests/test_memcache.py]: I guess this isn't accomplishing much.

File: tests/test_memcache.py [hunk 7]
```diff
@@ -137,6 +138,37 @@ def test_unicode_key(self):
...
+        self.assertEqual(compressed_value, compressed_result)
+        self.assertEqual(value, decompress(compressed_result).decode())
+
+        # Test add
+        self.mc.add('binary1-add', compressed_value)
+        compressed_result = self.mc.get('binary1-add')
+        self.assertEqual(compressed_value, compressed_result)
+        self.assertEqual(value, decompress(compressed_result).decode())
+
+        # Test set_many
```
Comment (timgraham) [inline tests/test_memcache.py]: The comments don't seem to add value (reviewing my own code). ;-)

File: tests/test_memcache.py [hunk 8]
```diff
@@ -137,6 +138,37 @@ def test_unicode_key(self):
...
+        compressed_value = compress(value.encode())
+
+        # Test set
+        self.mc.set('binary1', compressed_value)
+        compressed_result = self.mc.get('binary1')
+        self.assertEqual(compressed_value, compressed_result)
+        self.assertEqual(value, decompress(compressed_result).decode())
+
+        # Test add
+        self.mc.add('binary1-add', compressed_value)
```
Comment (timgraham) [inline tests/test_memcache.py]: Now that I see the fix, I wonder if the 3 variants are needed if they all use the same encoding process? For example, we don't have 3 variants for `test_unicode_value`.

File: tests/test_memcache.py (hunk 5 above)
Comment (adamchainz) [inline tests/test_memcache.py]: +1

File: memcache.py [hunk 9]
```diff
@@ -1266,11 +1272,10 @@ def _recv_value(self, server, flags, rlen):
...
         if flags == 0:
-            # Bare string
-            if six.PY3:
-                val = buf.decode('utf8')
-            else:
-                val = buf
+            # Bare bytes
+            val = buf
+        elif flags & Client._FLAG_TEXT:
+            val = buf.decode('utf8')
```
Comment (adamchainz) [inline memcache.py]: whilst you're at it you could make this `utf-8` to be consistent with line 980

**Comment by timgraham** [comment]
It's unclear to me if any of the `instance` -> `type` changes are required and completely tested. That commit was separated nicely in https://github.com/linsomniac/python-memcached/pull/74 although since that PR doesn't have tests it's not clear what is fixing what.

File: tests/test_memcache.py (hunk 8 above)
Comment (niconoe) [inline tests/test_memcache.py]: Isn't that more future-proof to ensure it works with the 3 methods. In that case, I'll complete `test_unicode_value()` instead.

File: memcache.py [hunk 10]
```diff
@@ -243,9 +244,9 @@ def __init__(self, servers, debug=0, pickleProtocol=0,
     def _encode_key(self, key):
         if isinstance(key, tuple):
             if isinstance(key[1], six.text_type):
-                return (key[0], key[1].encode('utf8'))
+                return (key[0], key[1].encode('utf-8'))
```
Comment (timgraham) [inline memcache.py]: Please don't make unrelated cleanups like that in a bug fix branch as it makes review more difficult, do it in a separate PR. (You've accidentally merged in the `touch()` fixes on this branch as well.

**Comment by timgraham** [comment]
@niconoe, do you have time to update this and remove the unrelated changes? I'd rebase the branch and make the fix a single squashed commit.

**Comment by niconoe** [comment]
@timgraham like this?

**Comment by timgraham** [comment]
Updated and merged in https://github.com/linsomniac/python-memcached/pull/135, thanks!
